README.md
# Pact Consumer Javascript DSL

[![Build Status](https://travis-ci.org/DiUS/pact-consumer-js-dsl.svg)](https://travis-ci.org/DiUS/pact-consumer-js-dsl)

This codebase provides a Javascript DSL for creating pacts. If you are new to Pact, please read the Pact [README][pact-readme] first.

The Javascript DSL is compatible with v2 of the [pact-specification](https://github.com/bethesque/pact-specification/tree/version-2) and supports type based matching, flexible array lengths, and regular expressions (read more below).

This DSL relies on the Ruby [pact-mock_service][pact-mock-service] gem to provide the mock service for the Javascript tests. If you do not want to use Ruby in your project, please read about using a standalone Pact mock service [here][pact-mock-service-without-ruby].

### Getting Help

* Please see the [wiki](https://github.com/DiUS/pact-consumer-js-dsl/wiki) for documentation (eg. using CORS, using the DSL on Windows).
* Google users group: https://groups.google.com/forum/#!forum/pact-support
* Twitter: [@pact_up](https://twitter.com/pact_up)

### Installing pact-mock-service

* If on Windows, please refer to the [Installing pact-mock-service on Windows](https://github.com/bethesque/pact-mock_service/wiki/Installing-the-pact-mock_service-gem-on-Windows)

1. You must install [Ruby](https://www.ruby-lang.org/en/downloads/) and [RubyGems](https://rubygems.org/pages/download) first.

1. After these binaries are available in the console, you can install the mock service by creating a Gemfile as shown below (this is the Ruby equivalent of package.json), then running `gem install bundler && bundle install` (the equivalent of npm install).

* Note: Windows users must run the install command after following Wiki instructions

```ruby
source 'https://rubygems.org'
gem 'pact-mock_service', '~> 0.7.0'

```

### Getting Started (with Karma, Jasmine and pact-mock-service)

1. Install and configure Karma with Jasmine

  1. Create a `package.json` if you don't have one already - use `npm init` if you don't

  1. Install Karma using their [installation instructions](http://karma-runner.github.io/0.12/intro/installation.html)

    This basically consists of running,

    ```
    npm install karma karma-jasmine karma-chrome-launcher --save-dev
    npm install -g karma-cli
    ```

  1. Initialise and configure Karma

    Run `karma init`. Answer **jasmine** for *testing framework* and **no** for *use require.js*.

  1. Add `pact-consumer-js-dsl` to your project by running `bower install pact-consumer-js-dsl --save-dev`.

  1. Tell Karma about `pact-consumer-js-dsl.js` in `karma.conf.js`. In the `files: []` section add a new entry for `bower_components/pact-consumer-js-dsl/dist/web/pact-consumer-js-dsl.js`.

  1. Allow tests to load resources from `pact` mock server. One way to do this is in the `karma.conf.js`, change `browsers: ['Chrome'],` or `browsers: ['PhantomJS'],` to,

         ````javascript
         browsers: ['Chrome_without_security'],
         customLaunchers: {
            Chrome_without_security: {
                base: 'Chrome',
                flags: ['--disable-web-security']
            }
         }

         or:

         browsers: ['PhantomJS_without_security'],
         customLaunchers: {
            PhantomJS_without_security: {
              base: 'PhantomJS',
              flags: ['--web-security=false']
            }
         }
         ````

   Note that running your tests across multiple browsers with one pact mock server will probably conflict with each other. You will need to either run them sequentially or start multiple pact mock servers. To run them sequentially make multiple calls to karma from the command line with the different browsers passed with the `--browser` option.

1. Write a Jasmine unit test similar to the following,

    ```javascript
    describe("Client", function() {
      var client, helloProvider;

      beforeAll(function(done) {
        //ProviderClient is the class you have written to make the HTTP calls to the provider
        client = new ProviderClient('http://localhost:1234');
        helloProvider = Pact.mockService({
          consumer: 'Hello Consumer',
          provider: 'Hello Provider',
          port: 1234,
          done: function (error) {
            expect(error).toBe(null);
          }
        });

        // This ensures your pact-mock-service is in a clean state before
        // running your test suite.
        helloProvider.resetSession(done);
      });

      it("should say hello", function(done) {
        helloProvider
          .given("an alligator with the name Mary exists")
          .uponReceiving("a request for an alligator")
          .withRequest("get", "/alligators/Mary", {
            "Accept": "application/json"
          }).willRespondWith(200, {
            "Content-Type": "application/json"
          }, {
            "name": "Mary"
          });

        helloProvider.run(done, function(runComplete) {
          expect(client.getAlligatorByName("Mary")).toEqual(new Alligator("Mary"));
          runComplete();
        });
      });
    });
    ```

    The "done" callback is used by the pact framework to communicate to your test framework that the expected interactions have not occurred. It should contain an assertion that will fail the test if an error is present. eg. for Jasmine:

    ```javascript
    done: function (error) {
      expect(error).toBe(null);
    }
    ```
    This is required because of the asynchonous nature of Javascript - raising exceptions to fail a test is not the "Javascript Way".

    See the spec in the example directory for more examples of asynchronous callbacks, how to expect error responses, and how to use query params.

    Make sure the source and test files are included by Karma in the `karma.conf.js` in the files array.

1. Let's run that bad boy!

   * Start the pact mock server with `bundle exec pact-mock-service -p 1234 --pact-specification-version 2.0.0 -l log/pact.logs --pact-dir tmp/pacts`
   * Run `karma start` (in another terminal window)
   * Inspect the pact file that has been written to "hello_consumer-hello_provider.json"

### Flexible Matching

Please read about using regular expressions and type based matching [here][flexible-matching] before continuing.

#### Match by regular expression

Remember that the mock service is written in Ruby, so the regular expression must be in a Ruby format, not a Javascript format. Make sure to start the mock service with the argument `--pact-specification-version 2.0.0`.

```javascript

provider
  .given('there is a product')
  .uponReceiving("request for products")
  .withRequest({
    method: "get",
    path: "/products",
    query: {
      category: Pact.Match.term({matcher: "\\w+", generate: 'pizza'}),
    }
  })
  .willRespondWith(
    200,
    {},
    {
      "collection": [
        {
          guid: Pact.Match.term({matcher: "\\d{16}", generate: "1111222233334444"})
        }
      ]
    }
  );
```

#### Match based on type

```javascript

provider
  .given('there is a product')
  .uponReceiving("request for products")
  .withRequest({
    method: "get",
    path: "/products",
    query: {
      category: Pact.Match.somethingLike("pizza")
    }
  })
  .willRespondWith(
    200,
    {},
    {
      "collection": [
        {
          guid: Pact.Match.somethingLike(1111222233334444)
        }
      ]
    }
  );
```

[flexible-matching]: https://github.com/realestate-com-au/pact/wiki/Regular-expressions-and-type-matching-with-Pact

#### Match based on arrays

Matching provides the ability to specify flexible length arrays. For example:

```javascript
Pact.Match.eachLike(obj, { min: 3 })
```

Where `obj` can be any javascript object, value or Pact.Match. It takes optional argument (`{ min: 3 }`) where min is greater than 0 and defaults to 1 if not provided. 

Below is an example that uses all of the Pact Matchers.

```javascript

var somethingLike = Pact.Match.somethingLike;
var term = Pact.Match.term;
var eachLike = Pact.Match.eachLike;

provider
  .given('there is a product')
  .uponReceiving("request for products")
  .withRequest({
    method: "get",
    path: "/products",
    query: {
      category: "clothing"
    }
  })
  .willRespondWith({
    status: 200,
    headers: {
        'Content-Type': 'application/json'
    },
    body: {
        "items":eachLike({
            size: somethingLike(10),
            colour: term("red|green|blue", {generates: "blue"}),
            tag: eachLike(somethingLike("jumper"))
        }, {min: 2})
    }
  });
```

### Examples

#### Web Example

Have a look at the [example](/example/web) folder. Ensure you have Google Chrome installed.

    $ cd example/web
    $ bundle install
    $ npm install
    $ npm test


#### Nodejs Example

This is only an example on how to use the pact-consumer-js-dsl within Node.  This is not best practice, but is a good starting point without creating a lengthy example using Grunt or Gulp with Protractor or Karma.

1. Start pact-mock-service with `bundle exec pact-mock-service start -p 1234 -l tmp/pact.log --pact-dir tmp/pacts`

1. Run nodejs command to setup pact consumer `node example/nodejs/setup.js`

1. Run your tests here with whatever you want, like Protractor for e2e testing

1. Run nodejs command to verify interactions and write pact files `node example/nodejs/teardown.js`

1. Stop the mock service with `bundle exec pact-mock-service stop -p 1234`

# Contributing

Please read [CONTRIBUTING.md](/CONTRIBUTING.md)

[pact-readme]: https://github.com/realestate-com-au/pact
[releases]: https://github.com/DiUS/pact-consumer-js-dsl/releases
[pact-mock-service]: https://github.com/bethesque/pact-mock_service
[pact-mock-service-without-ruby]: https://github.com/DiUS/pact-consumer-js-dsl/wiki/Using-the-Pact-Mock-Service-without-Ruby


CHANGELOG.md
# Changelog

Do this to generate your change history

  git log --pretty=format:'  * %h - %s (%an, %ad)'

### 0.2.2 (8 May 2016)

  * cf96a12 - Bump Pact Mock Service version in example for flexible matching (Matt Fellows, Sat May 7 22:17:03 2016 +1000)
  * 137d802 - Merge pull request #54 from aaronrenner/ar-reset-session (Matt Fellows, Sat May 7 21:41:09 2016 +1000)
  * f28cbb5 - Added resetSession to allow clearing out pact-mock-server (Aaron Renner, Thu Apr 21 13:25:52 2016 -0600)

### 0.2.1 (28 Sep 2015)

* 5794963 - Add eachLike array flexible matcher (Michael Isgro, Fri Sep 4 15:07:15 2015 +1000)

### 0.2 (7 Sep 2015)

* 88389ba - Added Pact Term and and SomethingLike flexible matchers. (Matt Fellows, Thu Aug 27 12:11:06 2015 +1000)
* ea11005 - set travis to run in docker containers (Ronald Holshausen, Tue Aug 25 20:15:53 2015 +1000)
* a33eb54 - update travis to use node as the language (Ronald Holshausen, Tue Aug 25 19:32:52 2015 +1000)
* 7c6434f - added travis build status badge (Ronald Holshausen, Tue Aug 25 19:17:46 2015 +1000)
* 9442f36 - Unit tests for the MockService configuration (Matt Fellows, Tue Aug 25 09:01:14 2015 +1000)
* 4e0f0be - Allow mock service host to be configured (Andrew Browne, Tue Apr 21 10:57:59 2015 +1000)
* 9760b87 - Updating example with the given method as it is required to add provider_state (cwmyers, Fri Feb 27 13:48:15 2015 +1100)
* 6428ad8 - Using PUT /interactions instead of DELETE /interactions and POST /interactions (Beth, Wed Feb 25 17:42:36 2015 +1100)
* e465e88 - Bump pact-mock_service gem version (Beth, Wed Feb 25 14:08:48 2015 +1100)

### 0.1.3 (25 Feb 2015)

* 0debea5 - Fixing examples. (Beth Skurrie, Wed Feb 25 14:02:46 2015 +1100)
* db5bd4e - removing custom UMD template, adding scoped XMLHttpRequest within Pact.Http, fixing tests accordingly (Michel Boudreau, Tue Feb 24 10:44:14 2015 +1100)
* eadd791 - adding run-sequence to package.json (Michel Boudreau, Tue Feb 24 09:11:49 2015 +1100)
* b1c5324 - fixing tests to be more agnostic, changing gulpfile to work better sequencially (Michel Boudreau, Mon Feb 23 17:27:53 2015 +1100)
* cfd60c2 - fixing gulp and synchronous tasks so there's no conflict in pact, added some code at top of tests to run in both web and node mode (Michel Boudreau, Mon Feb 23 16:57:59 2015 +1100)
* 525c762 - Merging in @mboudreau's changes to the MockService and adding unit tests. (Beth Skurrie, Thu Feb 19 12:57:27 2015 +1100)
* 5690bed - Merging in packaging and gulp build changes from @mboudreau (Beth Skurrie, Wed Feb 18 08:48:19 2015 +1100)
* 6c372c4 - Fixing typos and improving laying in README. Putting bundle exec back in. (Beth Skurrie, Wed Feb 18 08:47:31 2015 +1100)
* afbee20 - Merging documentation changes from @mboudreau (Beth Skurrie, Wed Feb 18 08:46:36 2015 +1100)

### 0.1.2 (25 Feb 2015)

* feec9cc - Updated CONTRIBUTING.md with npm publish instructions (Beth, Wed Feb 25 08:50:21 2015 +1100)
* c46ed6e - DRY up mock service response handling code. Raise error if 'done' option not provided. (Beth, Tue Feb 3 20:59:42 2015 +1100)
* e71bc02 - Accept fail test callback via constructor for mock service to allow it to be set in a beforeEach (Ben Sayers, Mon Feb 2 19:30:57 2015 +1100)
* 414e101 - Add an initial wait for the pact server to spin up so the tests don't execute too soon. (Ben Sayers, Thu Jan 22 13:03:32 2015 +1100)
* 5376021 - Add done callback to mockService to allow errors to be communicated back to the tests in a async friendly way (Ben Sayers, Thu Jan 15 22:39:15 2015 +1100)
* 019bb15 - Update run-tests task to return non-0 status code when the pact server fails to start (Ben Sayers, Tue Jan 13 23:08:11 2015 +1100)
* ecf8750 - Fix tests to fail when pact verification fails (Ben Sayers, Tue Jan 13 23:04:08 2015 +1100)
* 0be652d - Update package.json to point to correct main file and update docs with latest details on how to run the project (Ben Sayers, Tue Jan 13 23:02:31 2015 +1100)
* 96cacaf - Fixed #15 - Add nodejs support (Ben Sayers, Tue Jan 6 22:38:06 2015 +1100)
* 66a501c - Update mockService to make asynchronous http requests (Ben Sayers, Tue Dec 23 23:21:40 2014 +1100)
* b4a47d7 - Moved pact dir config to mock service command line (Beth Skurrie, Wed Jan 28 17:23:02 2015 +1100)
* bca90f8 - Update README.md (Beth Skurrie, Wed Jan 21 09:31:04 2015 +1100)
* dfb06fe - Fixed https://github.com/bethesque/pact-mock_service/issues/9 It should actually be upper case, but we need to fix https://github.com/bethesque/pact-support/issues/3 first (Beth, Fri Jan 16 18:11:46 2015 +1100)
* 91b1049 - Updated package.json (Beth, Fri Jan 9 11:15:59 2015 +1100)

### 0.1.0 (3 Feb 2015)

* e71bc02 - Accept fail test callback via constructor for mock service to allow it to be set in a beforeEach (Ben Sayers, Mon Feb 2 19:30:57 2015 +1100)
* 414e101 - Add an initial wait for the pact server to spin up so the tests don't execute too soon. (Ben Sayers, Thu Jan 22 13:03:32 2015 +1100)
* 5376021 - Add done callback to mockService to allow errors to be communicated back to the tests in a async friendly way (Ben Sayers, Thu Jan 15 22:39:15 2015 +1100)
* 019bb15 - Update run-tests task to return non-0 status code when the pact server fails to start (Ben Sayers, Tue Jan 13 23:08:11 2015 +1100)
* ecf8750 - Fix tests to fail when pact verification fails (Ben Sayers, Tue Jan 13 23:04:08 2015 +1100)
* 0be652d - Update package.json to point to correct main file and update docs with latest details on how to run the project (Ben Sayers, Tue Jan 13 23:02:31 2015 +1100)
* 96cacaf - Fixed #15 - Add nodejs support (Ben Sayers, Tue Jan 6 22:38:06 2015 +1100)
* 66a501c - Update mockService to make asynchronous http requests (Ben Sayers, Tue Dec 23 23:21:40 2014 +1100)
* b4a47d7 - Moved pact dir config to mock service command line (Beth Skurrie, Wed Jan 28 17:23:02 2015 +1100)

gulpfile.js
'use strict';

var gulp = require('gulp'),
	runSequence = require('run-sequence'),
    fs = require('fs-extra'),
    q = require('q'),
    request = require('request'),
    spawn = require('child_process').spawn,
    gutil = require('gulp-util'),
    umd = require('gulp-umd'),
    path = require('path'),
    karma = require('karma').server,
    $ = require('gulp-load-plugins')();

// FILES
var srcFiles = ['src/pact.js', 'src/interaction.js', 'src/http.js', 'src/mockServiceRequests.js', 'src/mockService.js', 'src/match.js'];
var specFiles = ['spec/**/*helper*.js', 'spec/**/*spec.js'];
var intSpecFiles = ['spec/integration/*spec.js', 'spec/**/*helper*.js'];
var intSpecPact2Files = ['spec/integration/**/*.js', 'spec/**/*helper*.js'];
var distFiles = ['dist/pact-consumer-js-dsl.js'];
var karmaConfig = 'spec/karma.conf.js';
var nodeConfig = 'spec/node.conf.js';

var cleanDirectories = function (directories) {
    directories.forEach(function (directory) {
        fs.removeSync(directory);
        fs.mkdirsSync(directory);
    });
};

var waitForServerToStart = function () {
    var deferred = q.defer();
    var attempts = 0;

    var checkIfServerHasStarted = function () {
        attempts += 1;
        request('http://localhost:1234', function (error) {
            if (attempts > 500) {
	            var e = new Error('Timed out waiting for the pact-mock-service to start');
	            console.error(e);
                deferred.reject(e);
            } else if (error) {
                setTimeout(checkIfServerHasStarted, 1000);
            } else {
                deferred.resolve();
            }
        });
    };

    setTimeout(checkIfServerHasStarted, 1000);

    return deferred.promise;
};

var withServer = function (action, version) {
    var deferred = q.defer();
    var pactSpecificationVersionArgs = [];

    if (!version) {
        version = 1;
    } else {
        pactSpecificationVersionArgs = ['--pact-specification-version', version];
    }

    cleanDirectories(['tmp/pacts', 'log']);
    var child = spawn('bundle', ['exec', 'pact-mock-service', '-p', '1234', '-l', 'tmp/pact.log', '--pact-dir', './tmp/pacts'].concat(pactSpecificationVersionArgs));
    child.on('error', function (error) {
        console.log('pact-mock-service:', error.toString());
    });

    waitForServerToStart().then(function () {
        var actionDone = function (error) {
            child.kill();

            if (error) {
                deferred.reject(error);
            } else {
                deferred.resolve();
            }
        };

        var actionStream = action();
        actionStream.on('end', actionDone);
        actionStream.on('error', actionDone);
        actionStream.resume();
    }, deferred.reject);

    return deferred.promise;
};

gulp.task('clear', function (done) {
    return $.cache.clearAll(done);
});

gulp.task('clean', ['clear'], function () {
    return gulp.src(['dist'], {
        read: false
    }).pipe($.clean());
});

gulp.task('build', ['clean'], function () {
    return gulp.src(srcFiles)
        .pipe($.jshint())
        .pipe($.jshint.reporter('default'))
        .pipe($.concat('pact-consumer-js-dsl.js'))
        .pipe(umd({
            exports: function (file) {
                return 'Pact';
            },
            namespace: function (file) {
                return 'Pact';
            }
        }))
        .pipe(gulp.dest('dist'))
        .pipe($.size())
});

gulp.task('default', ['build', 'run-tests']);

gulp.task('run-browser-tests', ['build'], function () {
    return withServer(function () {
        return gulp.src(distFiles.concat(intSpecFiles))
            .pipe($.karma({configFile: karmaConfig}));
    }, 1);
});

gulp.task('run-browser-tests-pact2', ['build'], function () {
    // NOTE: This is a superset of run-browser-tests, executed
    //       against a v2 version of the Pact Specification
    return withServer(function () {
        return gulp.src(distFiles.concat(intSpecPact2Files))
		    .pipe($.karma({configFile: 'spec/karma-v2.conf.js'}));
    }, 2);
});

gulp.task('run-unit-tests', ['build'], function () {
    return withServer(function () {
	    return gulp.src(distFiles.concat(['spec/unit/**/*spec.js']))
		    .pipe($.karma({configFile: 'spec/karma-unit.conf.js'}));
    });
});

gulp.task('run-node-tests', ['build'], function () {
    return withServer(function () {
        return gulp.src(distFiles.concat(nodeConfig, specFiles))
            .pipe($.jasmine());
    });
});

gulp.task('run-integration-tests', function(callback){
    runSequence('run-browser-tests-pact2', 'run-browser-tests', callback);
});

gulp.task('run-tests', function(callback){
	runSequence('run-unit-tests', 'run-browser-tests', 'run-browser-tests-pact2', 'run-node-tests', callback);
});

gulp.task('watch', ['clean'], function () {
    return withServer(function () {
        return gulp.src(srcFiles.concat(specFiles))
            .pipe($.karma({configFile: karmaConfig, action: 'watch'}));
    });
});

